Makes hacking use the electrical skill.

afterthought2 opened the pull request:
:cl:
rscadd: Hacking without the electrical skill is harder now.
rscadd: If you are unskilled in electrical, hacking will randomize the wire positions (not meanings) each time you open the screen. If you are basic, you'll get some possible duplicate colors/omitted colors, with smaller probabilities.
rscadd: If you pulse a wire below trained skill, you have (progressively smaller) chances of pulsing the wrong wire. If you are untrained, you may randomize wire positions for everyone else trying to hack the thing.
rscadd: If you cut/mend wires below trained skill, you may cut/mend some extra wires by accident.
rscadd: If you try to place a remote signaler with low skill (trained or lower), you might place it on the wrong wire, and then have a hard time removing it.
rscadd: A new "Examine" option has been added to the hacking menu. Below experienced it doesn't do much. On experienced one or two wires will be identifiable on examine, while at master most will be. Less effective for "secure" objects with randomized wiring.
/:cl:

afterthought2 committed:
Makes hacking use the electrical skill.

Sbotkin commented:
>If you pulse a wire below master skill, you have (progressively smaller) chances of also pulsing extra wires (very low on experienced).

This is too much imo. You need to be a little bit retarded to pulse two separate wires simultaneously, even trained wouldn't do this.

afterthought2 commented:
It's something like 4% chance on experienced, 8% on trained. You'll be doing it pretty rarely.

I have a problem that higher skill levels for this skill don't have much to differentiate themselves by, hence what you are seeing here. If people have further ideas that aren't too crazy to code, let me know (probably on the discord, not here).

Sbotkin commented:
Why not make it possible to see what the wire does on master?

Alex6511 commented:
I don't even understand how you could possibly connect the multi tool to to more than one wire at a time, generally they only have two connectors and both would have to be connected to the same wire to do anything.

afterthought2 committed:
Reduces RNG, gives an examine option.

afterthought2 commented:
Modified to account for feedback, and fixed some code issues.

Changes:

- Added an examine option that will let you see what the wires do on Master (and a bit on Experienced).

- Made pulsing pulse the wrong wire instead of an extra wire (though you'll sometimes pulse an extra one too)

- Capped the pulse wrong wire chance to below Trained.

Not entirely happy with it, especially around the Trained-Experienced mark, but that's all I have at the moment.

Alex6511 commented:
This will require wiki changes at merge-time, as well as changes to the SM book. The current guide encourages counting shots.

PsiOmegaDelta commented on code/datums/wires/airlock.dm:
```diff
@@ -174,3 +174,26 @@ var/const/AIRLOCK_WIRE_LIGHT = 2048
 			A.update_icon()
+
+/datum/wires/airlock/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/alarm.dm:
```diff
@@ -100,3 +100,20 @@ var/const/AALARM_WIRE_AALARM = 16
 			A.update_icon()
+
+/datum/wires/alarm/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/apc.dm:
```diff
@@ -70,3 +70,16 @@
 					A.aidisabled = 0
+
+/datum/wires/apc/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/airlock.dm:
```diff
@@ -174,3 +174,26 @@ var/const/AIRLOCK_WIRE_LIGHT = 2048
+
+/datum/wires/airlock/examine(index, mob/user)
+	. = ..()
+	if(user.skill_check(SKILL_ELECTRICAL, SKILL_EXPERT))
```
I'll note that this will basically invalidate any and all secure wires that randomize their setup.
No matter your expertise at some point it doesn't seem reasonable to be able to tell what a random wire does just by looking at it.

Sbotkin replied:
tbh you should be able to recognize what a wire do if you are an experienced electrician. Otherwise it makes little sense.

afterthought2 replied:
The whole system is unreasonable, before or after the PR. From a gameplay perspective, I don't see an issue here. From a realism perspective, it gives a clear mechanism by which this works--you follow the wire to see where it goes, or measure the current--that doesn't vary with whether or not the object is secure. The way secure versions work is, if anything, not realistic.

babydoll replied:
It doesn't matter what is or isn't realistic, the purpose of the secure wires is to hinder people hacking them

afterthought2 replied:
If that's the issue here, I stand by this. This means you took master in electrical (as opposed to, say, a combat skill) as an antag, and deserve to reap some rewards. Hacking a few more airlocks a bit faster seems balanced to me. On experienced, the examine ability isn't that good, especially on airlocks.

Sbotkin replied:
Make it only for master in case of secure airlocks then.

PsiOmegaDelta replied:
If we're talking realism I view airlock electronics to be more of a control board than something more overview such as Arduinos. The only reason it's actual wires is more due to fact that that's the hacking system we have.

And if you wish to bring up antags this will make it easier than ever to counter Malfs if you have engineers, or similar staff around, because now secure airlocks will give them even less of a breathing room (along with all other defense issues they have).

PsiOmegaDelta commented on code/datums/wires/autolathe.dm:
```diff
@@ -58,3 +58,16 @@ var/const/AUTOLATHE_DISABLE_WIRE = 4
 					Interact(usr)
+
+/datum/wires/autolathe/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/camera.dm:
```diff
@@ -69,3 +69,18 @@ var/const/CAMERA_WIRE_NOTHING2 = 32
 		return 0
+
+/datum/wires/camera/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/nuclearbomb.dm:
```diff
@@ -56,3 +56,16 @@ var/const/NUCLEARBOMB_WIRE_SAFETY		= 4
 			N.update_icon()
+
+/datum/wires/nuclearbomb/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/particle_accelerator.dm:
```diff
@@ -50,3 +50,18 @@ var/const/PARTICLE_LIMIT_POWER_WIRE = 8 // Determines how strong the PA can be.
 				C.remove_strength()
+
+/datum/wires/particle_acc/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/radio.dm:
```diff
@@ -45,3 +45,16 @@ var/const/WIRE_TRANSMIT = 4
 			R.broadcasting = mended && !IsIndexCut(WIRE_SIGNAL)
 	SSnano.update_uis(holder)
+
+/datum/wires/radio/examine(index, mob/user)
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/robot.dm:
```diff
@@ -75,3 +71,20 @@ var/const/BORG_WIRE_CAMERA = 16
 	return wires_status & BORG_WIRE_LOCKED_DOWN
+
+/datum/wires/robot/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/shield_generator.dm:
```diff
@@ -37,10 +37,25 @@ var/const/SHIELDGEN_WIRE_NOTHING = 16		// A blank wire that doesn't have any spe
+	. += "<BR>A yellow light labeled \"Interface Connection\" is [S.mode_changes_locked ? "off" : "on"].<BR>"
+
+/datum/wires/shield_generator/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/smartfridge.dm:
```diff
@@ -49,3 +49,16 @@ var/const/SMARTFRIDGE_WIRE_IDSCAN		= 4
 			S.scan_id = 1
+
+/datum/wires/smartfridge/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/smes.dm:
```diff
@@ -57,4 +57,21 @@ var/const/SMES_WIRE_FAILSAFES = 16	// Cut to disable failsafes, mend to reenable
+					S.safeties_enabled = 1
+
+/datum/wires/smes/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/suit_storage_unit.dm:
```diff
@@ -45,3 +45,16 @@ var/const/SUIT_STORAGE_WIRE_LOCKED		= 4
 				S.electrified = -1
+
+/datum/wires/suit_storage_unit/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/vending.dm:
```diff
@@ -53,3 +53,18 @@ var/const/VENDING_WIRE_IDSCAN = 8
 			V.scan_id = 1
+
+/datum/wires/vending/examine(index, mob/user)
+	. = ..()
```
```
if((. = ..()))
   return
```

PsiOmegaDelta commented on code/datums/wires/airlock.dm:
```diff
@@ -174,3 +174,26 @@ var/const/AIRLOCK_WIRE_LIGHT = 2048
+/datum/wires/airlock/examine(index, mob/user)
+	. = ..()
+	if(user.skill_check(SKILL_ELECTRICAL, SKILL_EXPERT))
+		switch(index)
```
This examination system could be fairly generic/extensible.

Instead if you want to add, for example, a check for secure wires you have to go around and copy-paste it in every single wire examination override.

afterthought2 replied:
I don't really know how to do that without making changes that are outside the scope of the PR (and outside the scope of the time I have to devote to this right now). I considered making wire effect datums that have procs to be cut, examined, etc, and are then called by the wire datum. Unfortunately this would require a lot of work, as the machinery effects these actions have are all machine-specific (little reusable code) and some machines check for specific wire state from outside the wire datum itself.

If you have advice for how to improve this, I'm open to it, but I can't bring myself to refactor the whole setup.

PsiOmegaDelta replied:
I don't particularly feel like approving what's basically copy-paste, whether or not you consider that to be in scope.

Below is a simple system that keeps functionality in one place, instead of having each and every subtype having to re-implement examination themselves:
```
/datum/wire_description
	var/index
	var/description
	var/skill_level // Optional with the propose setup

/datum/wire_description/New(var/index, var/description, var/skill_level)
	src.index = index
	src.description = description
	src.skill_level = skill_level
	
/datum/wires
	[...]
	var/list/descriptions
	
/datum/wires/New()
	..()
	var/descriptions_by_index[wire_count]
	for(var/datum/wire_description/wd in descriptions)
		descriptions_by_index[wd.index] = wd
	descriptions = descriptions_by_index	
	
/datum/wires/proc/examine(var/index, var/mob/user)
	. = "You aren't sure what this wire does."

	var/datum/wire_description/wd = descriptions[index]
	if(!wd)
		return 
	if(wd.skill_level && !user.skill_check(SKILL_ELECTRICAL, wd.skill_level))
		return
	return wd.description		
	
	
/datum/wires/subtype
	[...]
	descriptions = list(
		new/datum/wire_description(VENDING_WIRE_IDSCAN,     "Desc 1", SKILL_EXPERT), // I'm sure no one can figure out what this particular wire does using logical deduction
		new/datum/wire_description(VENDING_WIRE_THROW,      "Desc 2", SKILL_PROF),
		new/datum/wire_description(VENDING_WIRE_CONTRABAND, "Desc 3", SKILL_PROF)
		new/datum/wire_description(VENDING_WIRE_ELECTRIFY,  "Desc 4", SKILL_PROF)
	)
```

Bad-Apple797 commented:
I think this is great, exact balancing of chance or secure airlocks aside. (Personally I'd say half experience fail chance and half amount of known wires in a secure airlock and see how that plays)

Is it extensible for possible makeshift equipment in future (cut wire with shard, pulse wire with battery + cable coil) being possible with higher penalties?

Additionally to secure airlocks is there other equipment this should be adjusyed for (bombs, engine equipment)

Also how does it affect remote signallers?

afterthought2 committed:
Made descriptions into datums. Made secure stuff less likely to give descriptions.

afterthought2 commented:
All right, (finally) ready for further review. I made the descriptions into datums as per @PsiOmegaDelta's comment (thank you). The descriptions are not listed by index because the index is actually a flag here, and having a massive list with lots of nulls seems bad.

Also made secure stuff that randomizes wiring (this includes secure airlocks, secure smartfridges, all cameras, all borgs, and all rigs) have each wire require +1 skill with 50% probability to examine (this will, for most wires, make them impossible to get information from by examine, but will boost Expert skill wires to Master skill wires).

Sbotkin commented on code/datums/wires/airlock.dm:
```diff
@@ -9,6 +9,20 @@
 	window_y = 570
+	descriptions = list(
+		new /datum/wire_description(AIRLOCK_WIRE_IDSCAN, "This wire is connected to the ID scanning panel.", SKILL_EXPERT),
+		new /datum/wire_description(AIRLOCK_WIRE_MAIN_POWER1, "This wire seems to be carrying a heavy current."),
```
Isn't "current" uncountable?

afterthought2 replied:
You mean "a heavy current" vs. "heavy current"? I believe that it's correct as written, but possibly also valid without the article. It's not uncommon to make a distinction between the singular "a current" and plural "currents," even when referring specifically to an electrical current. I think omitting the article here would make it refer to current in the abstract (the concept of electrical current), as opposed to a particular current, which happens to have essentially the same meaning in the context.

comma commented:
IMO should keep Trained as 'working as now', without the duplicate etc chance.

afterthought2 committed:
Further waters down needed skills.
